zapcore/sampler_test.go

```diff
@@ -0,0 +1,168 @@
+	sampler, logs := fakeSampler(InfoLevel, time.Minute, 1, 100)
+
+	// Shouldn't be counted, because debug logging isn't enabled.
+	writeIter(sampler, 1, DebugLevel)
+	writeIter(sampler, 2, InfoLevel)
+	expected := buildExpectation(InfoLevel, 2)
+	assert.Equal(t, expected, logs.All(), "Expected to disregard disabled log levels.")
+}
+
+func TestSamplerWithSharesCounters(t *testing.T) {
+	sampler, logs := fakeSampler(DebugLevel, time.Minute, 1, 100)
+
+	first := sampler.With([]Field{makeInt64Field("child", 1)})
+	for i := 1; i < 10; i++ {
+		writeIter(first, i, InfoLevel)
+	}
+	second := sampler.With([]Field{makeInt64Field("child", 2)})
+	// The new child logger should share the same counters, so we don't expect to
+	// write these logs.
+	for i := 10; i < 20; i++ {
+		writeIter(second, i, InfoLevel)
+	}
+
+	expected := []observer.LoggedEntry{{
+		Entry:   Entry{Level: InfoLevel},
+		Context: []Field{makeInt64Field("child", 1), makeInt64Field("iter", 1)},
+	}}
+	assert.Equal(t, expected, logs.All(), "Expected child loggers to share counters.")
+}
+
+func TestSamplerTicks(t *testing.T) {
+	// Ensure that we're resetting the sampler's counter every tick.
+	sampler, logs := fakeSampler(DebugLevel, time.Millisecond, 1, 1000)
+
+	// The first statement should be logged, the second should be skipped but
+	// start the reset timer, and then we sleep. After sleeping for more than a
+	// tick, the third statement should be logged.
+	for i := 1; i < 4; i++ {
+		if i == 3 {
+			testutils.Sleep(5 * time.Millisecond)
+		}
+		writeIter(sampler, i, InfoLevel)
+	}
+
+	expected := buildExpectation(InfoLevel, 1, 3)
+	assert.Equal(t, expected, logs.All(), "Expected sleeping for a tick to reset sampler.")
+}
+
+func TestSamplerCheck(t *testing.T) {
+	sampler, logs := fakeSampler(InfoLevel, time.Millisecond, 1, 10)
+
+	assert.Nil(t, sampler.Check(Entry{Level: DebugLevel}, nil), "Expected a nil CheckedMessage at disabled log levels.")
+
+	for i := 1; i < 12; i++ {
+		if cm := sampler.Check(Entry{Level: InfoLevel}, nil); cm != nil {
+			cm.Write(makeInt64Field("iter", i))
+		}
+	}
+
+	expected := buildExpectation(InfoLevel, 1, 11)
+	assert.Equal(t, expected, logs.All(), "Unexpected output when sampling with Check.")
+}
+
+// TODO: restore this test, now that panic and fatal are actually terminal
+// func TestSamplerCheckPanicFatal(t *testing.T) {
+// 	for _, level := range []zap.Level{zap.FatalLevel, zap.PanicLevel} {
+// 		sampler, sink := fakeSampler(zap.FatalLevel+1, time.Millisecond, 1, 10, false)
+
+// 		assert.Nil(t, sampler.Check(zap.DebugLevel, "foo"), "Expected a nil CheckedMessage at disabled log levels.")
+// 		for i := 0; i < 5; i++ {
+// 			if cm := sampler.Check(level, "sample"); assert.True(t, cm.OK(), "expected %v level to always be OK", level) {
+// 				cm.Write(zap.Int("iter", i))
+// 			}
+// 		}
+
+// 		assert.Equal(t, []spy.Log(nil), sink.Logs(), "Unexpected output when sampling with Check.")
+// 	}
+// }
```

Comment:
Can we restore this test now? We should be able to use the internal `exit` package to make sure that we're actually sampling messages at panic and fatal too.